packages/firestore/test/integration/api/transactions.test.ts

```diff
@@ -63,131 +240,261 @@ apiDescribe('Database transactions', (persistence: boolean) => {
-            transaction.update(docRef, { foo: 'bar' });
-          });
-        })
-        .then(() => expect.fail('transaction should fail'))
-        .catch((err: firestore.FirestoreError) => {
-          expect(err).to.exist;
-          expect(err.code).to.equal('invalid-argument');
-        });
+  it('run transactions after getting existing document', async () => {
+    return integrationHelpers.withTestDb(persistence, async db => {
+      const tt = new TransactionTester(db);
+      await tt.validateSequenceSuccess(
+        undefined,
+        /** docExists = */ true,
+        tt.getExisting,
+        tt.delete,
+        tt.delete
+      );
+      await tt.validateSequenceError(
+        'invalid-argument',
+        /** docExists = */ true,
+        tt.getExisting,
+        tt.delete,
+        tt.update2
+      );
+      await tt.validateSequenceSuccess(
+        { foo: 'bar2' },
+        /** docExists = */ true,
+        tt.getExisting,
+        tt.delete,
+        tt.set2
+      );
+      await tt.validateSequenceSuccess(
+        undefined,
+        /** docExists = */ true,
+        tt.getExisting,
+        tt.update1,
+        tt.delete
+      );
+      await tt.validateSequenceSuccess(
+        { foo: 'bar2' },
+        /** docExists = */ true,
+        tt.getExisting,
+        tt.update1,
+        tt.update2
+      );
+      await tt.validateSequenceSuccess(
+        { foo: 'bar2' },
+        /** docExists = */ true,
+        tt.getExisting,
+        tt.update1,
+        tt.set2
+      );
+      await tt.validateSequenceSuccess(
+        undefined,
+        /** docExists = */ true,
+        tt.getExisting,
+        tt.set1,
+        tt.delete
+      );
+      await tt.validateSequenceSuccess(
+        { foo: 'bar2' },
+        /** docExists = */ true,
+        tt.getExisting,
+        tt.set1,
+        tt.update2
+      );
+      await tt.validateSequenceSuccess(
+        { foo: 'bar2' },
+        /** docExists = */ true,
+        tt.getExisting,
+        tt.set1,
+        tt.set2
+      );
     });
   });
 
-  it("can't delete document then patch", () => {
-    return integrationHelpers.withTestDb(persistence, db => {
-      const docRef = db.collection('towns').doc();
-      return docRef.set({ foo: 'bar' }).then(() => {
-        return db
-          .runTransaction(transaction => {
-            return transaction.get(docRef).then(docSnap => {
-              expect(docSnap.exists).to.equal(true);
-              transaction.delete(docRef);
-              // Since we deleted the doc, the update will fail
-              transaction.update(docRef, { foo: 'new-bar' });
-            });
-          })
-          .catch((err: firestore.FirestoreError) => {
-            expect(err).to.exist;
-            // This is the error surfaced by the backend.
-            expect(err.code).to.equal('invalid-argument');
-          });
-      });
+  it('run transactions after getting non-existent document', async () => {
```

Comment:
This test description is the same as above but docExists = false here.

Reply:
One gets an "existing" document; the other gets a "non-existent" document.